feat: functions requesting extra resources

phisco opened the pull request:
<!--
Thank you for helping to improve Crossplane! Please read the contribution docs
(linked below) if this is your first Crossplane pull request.
-->

### Description of your changes

<!--
Briefly describe what this pull request does, and how it is covered by tests.
Be proactive - direct your reviewers' attention to anything that needs special
consideration.

We love pull requests that fix an open issue. If yours does, use the below line
to indicate which issue it fixes, for example "Fixes #500".
-->

Implementation of https://github.com/crossplane/crossplane/pull/5099.

Tested out using https://github.com/phisco/function-environment-configs.

```sh
# after cloning https://github.com/phisco/function-environment-configs
$ cd examples
$ crank beta render xr.yaml composition.yaml functions.yaml -r --extra-resources environmentConfigs.yaml -c
---
apiVersion: example.crossplane.io/v1
kind: XR
metadata:
  name: example-xr
status:
  fromEnv: e
---
apiVersion: render.crossplane.io/v1beta1
fields:
  apiextensions.crossplane.io/environment:
    complex:
      a: b
      c:
        d: e
        f: "1"
kind: Context
```

The same example manifests can be used to deploy it locally using the image I pushed to `index.docker.io/phisco/function-environment-configs:v0.0.3`.

I have: <!--You MUST either [x] check or [ ] ~strike through~ every item.-->

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added or updated unit tests.
- [ ] ~Added or updated e2e tests.~ We'll add them later on once we've built a function leveraging the feature.
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR.~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

phisco committed:
feat: allow functions to request extra resources

This adds:
- the ability for functions to request extra resources.
- a flag to disable such functionality (default is enabled), resulting
  in any requirement being just ignored.
- two flags to the render command:
  - `--extra-resources` to specify extra resources that functions
    should be able to request.
  - `--include-context` to allow printing the context in the output
    of the render command.

Signed-off-by: Philippe Scorsolini <p.scorsolini@gmail.com>

negz commented on cmd/crank/beta/render/render.go:
```diff
@@ -257,3 +308,35 @@ func SetComposedResourceMetadata(cd resource.Object, xr resource.Composite, name
+	out := &fnv1beta1.Resources{}
+	for _, er := range ers {
+		er := er
+		if selector.GetApiVersion() != "" && selector.GetApiVersion() != er.GetAPIVersion() {
```
```suggestion
		if s := selector.GetApiVersion(); s != "" && s != er.GetAPIVersion() {
```

Nit: This could work too. Just a thought though, it's fine as is.

That said, why the check for `!= ""`? Do we expect to support empty API versions or kinds in the selector?

phisco replied:
no, we don't support empty API versions or kinds, it was indeed unnecessary.

negz commented on cmd/crank/beta/render/render.go:
```diff
@@ -150,24 +154,63 @@ func Render(ctx context.Context, in Inputs) (Outputs, error) { //nolint:gocyclo
+		var requirements *fnv1beta1.Requirements
+		var rsp *fnv1beta1.RunFunctionResponse
+
+		for i := int64(0); i < composite.MaxRequirementsIterations; i++ {
```
Is there anything you could move out of this inner loop? I'm seeing some things that I don't think need to happen for each (potential) re-run of a function:

* Unmarshaling the input.
* Getting a connection and creating a client.

I'd only prioritize this if it doesn't make the code flow awkward; not super worried about performance in the `render` loop. It probably does matter in the controller equivalent though.

negz commented on cmd/crank/beta/render/render.go:
```diff
@@ -150,24 +154,63 @@ func Render(ctx context.Context, in Inputs) (Outputs, error) { //nolint:gocyclo
+		// used to store the resources fetched at the previous iteration
+		var extraResources map[string]*fnv1beta1.Resources
+		// used to store the requirements returned at the previous iteration
+		var requirements *fnv1beta1.Requirements
```
Maybe `existing` or `previous` would make sense as a variable name? Could help make the `requirements = req` assignment below more self-documenting.

negz commented on cmd/crank/beta/render/render.go:
```diff
@@ -150,24 +154,63 @@ func Render(ctx context.Context, in Inputs) (Outputs, error) { //nolint:gocyclo
+
+			// if we reached the maximum number of iterations we need to return an error
+			if i == composite.MaxRequirementsIterations-1 {
+				return Outputs{}, errors.Errorf("pipeline step %q reached the maximum number of iterations (%d)", fn.Step, composite.MaxRequirementsIterations)
```
This error could be a little clearer as to why we reached maximum iterations - i.e. the requirements didn't stabilize. I'd also personally prefer/expect this "limit exceeded" check at the beginning of the loop.

phisco commented:
Except for the discussion going on at https://github.com/crossplane/function-sdk-go/pull/119#discussion_r1457876304, we can proceed reviewing and merging this PR, changes will be mostly a refactoring.

negz commented on internal/controller/apiextensions/composite/composition_functions.go:
```diff
@@ -104,6 +117,7 @@ type xr struct {
 	managed.ConnectionDetailsFetcher
 	ComposedResourceObserver
 	ComposedResourceGarbageCollector
+	ExtraResourcesGetter
```
Nit: Would `Fetcher` or `Observer` work here, to better match the two other interfaces that "get stuff"?

negz commented on internal/controller/apiextensions/composite/composition_functions.go:
```diff
@@ -133,6 +147,19 @@ func (fn ComposedResourceObserverFn) ObserveComposedResources(ctx context.Contex
 
+// A ExtraResourcesGetter gets extra resources matching a selector.
+type ExtraResourcesGetter interface {
+	Get(ctx context.Context, selector *v1beta1.ResourceSelector) (*v1beta1.Resources, error)
```
```suggestion
	Get(ctx context.Context, rs *v1beta1.ResourceSelector) (*v1beta1.Resources, error)
```

Nit: `selector` is redundant given the type.

negz commented on internal/controller/apiextensions/composite/composition_functions.go:
```diff
@@ -168,6 +195,14 @@ func WithComposedResourceObserver(g ComposedResourceObserver) FunctionComposerOp
 
+// WithExtraResourcesGetter configures how the FunctionComposer should fetch extra
+// resources requested by functions.
+func WithExtraResourcesGetter(f ExtraResourcesGetter) FunctionComposerOption {
```
Nit: Why `f`? I think it was probably short for `Fetcher`.

phisco replied:
yes, it was, it's now back to Fetcher though

negz commented on internal/controller/apiextensions/composite/composition_functions.go:
```diff
@@ -66,6 +71,7 @@ const (
 	errFmtUnmarshalDesiredCD         = "cannot unmarshal desired composed resource %q from RunFunctionResponse"
 	errFmtCDAsStruct                 = "cannot encode composed resource %q to protocol buffer Struct well-known type"
 	errFmtFatalResult                = "pipeline step %q returned a fatal result: %s"
+	errFmtFunctionMaxIterations      = "pipeline step %q reached the maximum number of iterations (%d)"
```
Similar to render, I think it would be useful to explain that we hit max iterations because the function's requirements didn't stabilize.

negz commented on internal/controller/apiextensions/composite/composition_functions.go:
```diff
@@ -91,6 +97,13 @@ const (
+const (
+	// MaxRequirementsIterations is the maximum number of times a Function should be called,
+	// limiting the number of times it can request for extra resources, capped for
+	// safety. We might allow Composition authors customize this value in the future.
```
```suggestion
	// safety.
```

Nit: I'd remove this from the type documentation. I think it's fine left unsaid, but you could add a distinct TODO comment nearby if you prefer.

phisco replied:
removed

negz commented on internal/controller/apiextensions/composite/composition_functions.go:
```diff
@@ -264,11 +299,56 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
+				return CompositionResult{}, errors.Wrapf(err, errFmtRunPipelineStep, fn.Step)
+			}
+
+			if c.composite.ExtraResourcesGetter == nil {
+				// If we don't have an extra resources getter, we don't need to
+				// iterate to satisfy the requirements.
+				break
+			}
```
I think this would only be true if the feature flag was disabled, right? If so, consider this a note that it could be removed if we decide to drop the distinct feature flag for extra resources.

phisco replied:
yes, only in that case indeed.

negz commented on internal/controller/apiextensions/composite/composition_functions.go:
```diff
@@ -264,11 +299,56 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
+		var rsp *v1beta1.RunFunctionResponse
+
+		for i := int64(0); i <= MaxRequirementsIterations; i++ {
+			// if we reached the maximum number of iterations we need to return an error
```
I'd maybe mention requirement stabilization in this comment too, otherwise its not saying much that the code doesn't already document pretty clearly.

negz commented on internal/controller/apiextensions/composite/composition_functions.go:
```diff
@@ -264,11 +299,56 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
+
+			// Fetch the requested resources and add them to the desired state.
+			for name, selector := range newRequirements.GetExtraResources() {
+				// Fetch the requested resources and add them to the desired state.
```
This is the same comment as two lines up. 🙂

negz commented on internal/controller/apiextensions/composite/composition_functions.go:
```diff
@@ -264,11 +299,56 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
-		if err != nil {
-			return CompositionResult{}, errors.Wrapf(err, errFmtRunPipelineStep, fn.Step)
+		// used to store the requirements returned at the previous iteration
+		var requirements *v1beta1.Requirements
```
```suggestion
		var previous *v1beta1.Requirements
```

Perhaps this would help document what it is? I can tell that it's Requirements from the type.

phisco replied:
we went with `d` and `fctx` for "storage" variables between steps, so I adopted the same pattern here. calling new the variables to store the temporary ones if needed.

negz commented on internal/controller/apiextensions/composite/composition_functions.go:
```diff
@@ -418,6 +498,66 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
+		err := e.client.Get(ctx, nn, r)
+		if kerrors.IsNotFound(err) {
+			// The resource doesn't exist. We'll return nil, which the Functions
+			// know it means that the resource was not found.
```
```suggestion
			// know means that the resource was not found.
```

negz reviewed (commented):
Pretty much all nits. This LGTM once updated per https://github.com/crossplane/function-sdk-go/pull/119#discussion_r1457876304.

phisco committed:
chore: review

1. Implemented changes for improved code semantics and naming conventions.
2. Enhanced readability through grammar correction in comments.
3. Improved user experience by refining the clarity of error messages.
4. Cleaned up the code by removing unnecessary comments and TODOs.

Signed-off-by: Philippe Scorsolini <p.scorsolini@gmail.com>

phisco commented on internal/features/features.go:
```diff
@@ -49,6 +49,11 @@ const (
 	// https://github.com/crossplane/crossplane/blob/863ff6/design/design-doc-composition-functions.md
 	EnableBetaCompositionFunctions feature.Flag = "EnableBetaCompositionFunctions"
 
+	// EnableBetaCompositionFunctionsExtraResources enables extra resources support for
+	// composition functions. See the below design for more details.
+	// https://github.com/crossplane/crossplane/blob/863ff6/design/design-doc-composition-functions.md
+	EnableBetaCompositionFunctionsExtraResources feature.Flag = "EnableBetaCompositionFunctionsExtraResources"
```
We were saying to keep a separate flag to allow disabling extra resource gathering for more security aware users, so I think it would make sense to keep it at least for now just to avoid having to add it back later as soon as someone asks for it.

negz replied:
Oh right, I forgot about that.

I'm not sure a feature flag is the best fit for it though, in the sense that I don't think of it as being a lifecycle/maturity thing. i.e. I don't think this is being a flag due to its maturity, but rather because folks might want to turn it off even after functions (and extra resources) are GA.

That said, using a feature flag is effectively an implementation detail for now, so we can worry about this later. I imagine you'd toggle it with `--enable-composition-functions-extra-resources` whether it was a feature flag (in the alpha/beta/GA lifecycle sense) or not.

phisco replied:
yes, exactly, I ended up using the feature flag because it was already wired properly and it was near to the other flag it is linked to, it's indeed `--enable-composition-functions-extra-resources` and once we promote to beta we can just change the implementation.

phisco commented on cmd/crank/beta/render/render.go:
```diff
@@ -160,14 +172,47 @@ func Render(ctx context.Context, in Inputs) (Outputs, error) { //nolint:gocyclo
+			for name, selector := range newRequirements.GetExtraResources() {
+				newExtraResources, err := filterExtraResources(in.ExtraResources, selector)
+				if err != nil {
+					return Outputs{}, errors.Wrapf(err /* TODO: */, "fetching resources for %s", name)
```
I think it was a reminder to update the error message, done.

phisco commented on cmd/crank/beta/render/render.go:
```diff
@@ -257,3 +310,38 @@ func SetComposedResourceMetadata(cd resource.Object, xr resource.Composite, name
 	return errors.Wrapf(meta.AddControllerReference(cd, or), "cannot set composite resource %q as controller ref of composed resource", xr.GetName())
 }
+
+func filterExtraResources(ers []unstructured.Unstructured, selector *fnv1beta1.ResourceSelector) (*fnv1beta1.Resources, error) { //nolint:gocyclo // TODO(phisco): refactor
```
no, updated the comment.

phisco commented on cmd/crank/beta/render/render.go:
```diff
@@ -150,6 +154,14 @@ func Render(ctx context.Context, in Inputs) (Outputs, error) { //nolint:gocyclo
+
+		fClient := fnv1beta1.NewFunctionRunnerServiceClient(conn)
+
+		// the request to send to the function, will be updated at each iteration if needed
```
done, both here and in the function composer.

negz reviewed (approved):
LGTM, pending addressing the nits. Thank you!

ezgidemirel reviewed (approved):
looks great, thanks @phisco!
